urle381b6cba7_IkarosKappler_ikrs_http_tgz-pJ8-ikrs_json_rpc_JSONRPCHandlerJ8/src/ikrs/httpd/resource/RangedResource.java

```diff
@@ -0,0 +1,203 @@
+     * @see isReadOnly()
+     **/
+    public void open( boolean readOnly )
+	throws IOException {
+
+	// Open the core resource first!
+	// (otherwise we will not be able to access the input stream)
+	super.open( readOnly );
+      
+	// No need to store the dat of last modification:
+	//  both resources share the same meta data instance.
+    }
+
+    /**
+     * Get the input stream from this resource.
+     *
+     * @throws IOException If any IO error occurs.
+     **/
+    public InputStream getInputStream()
+	throws IOException {
+
+	if( !this.isOpen() )
+	    throw new IOException( "Cannot retrieve resource's input stream. Resouce was not yet opened." );
+
+	if( this.readLimitInputStream == null ) {
+
+	    if( this.contentRange.getLastBytePosition() >= this.getCoreResource().getLength() )
+		throw new IOException( "Failed to initialize the RangedResource's input stream; the passed Content-Range " + this.contentRange.toString() + " is out of bounds!" );
+
+	    // Init the stream(s)!
+	    InputStream tmpIn = this.getCoreResource().getInputStream();
+	    if( this.contentRange.getFirstBytePosition() > 0 ) {
+		
+		// Skip n bytes at the beginning
+		tmpIn.skip( this.contentRange.getFirstBytePosition() );
+
+	    }
+
+	    this.readLimitInputStream = new ReadLimitInputStream( tmpIn,
+								  this.contentRange.calculateLength()
+								  );
+
+	}
+
+
+
+	return this.readLimitInputStream;
+    }
+
+    /**
+     * This method returns the *actual* length of the underlying resource. This length will
+     * be used in the HTTP header fields to specify the transaction length.
+     *
+     * During read-process (you used the locks, didn't you?) the length MUST NOT change.
+     *
+     * @return the length of the resource's data in bytes.
+     * @throws IOException If any IO error occurs.
+     **/
+    public long getLength()
+	throws IOException {
+
+	// The new length is now specified by the passed content range!
+	return this.contentRange.calculateLength();
+    }
+
+    //--- END ----------------------- Override ResourceDelegation -------------------------
+
+
+    /**
+     * This method is for testing only.
+     **/
+    public static void main( String[] argv ) {
+
+	try {
+
+	    System.out.println( "Creating file and content range ..." );
+	    java.io.File file = new java.io.File( "document_root/README.txt" );
+	    ContentRange range = new ContentRange( ContentRange.NAME_BYTESUNIT_BYTES,
+						   7L,   // skip the first 8 bytes ... (begin at index 7)
+						   15L,  // ... and read up to 8 bytes from the resource (end at index 15).
+						   -1    // unknown resource size
+						   );
+					     
+	    System.out.println( "Creating RangedResource ...." );
+	    Resource resource = new RangedResource( new FileResource( null,   // no handler available here
+								      null,   // and also no logger
+								      file,   // the input file
+								      false   // no need to use fair locks here
+								      ),
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
Analysis of this package determined that this line of code contains a resource that might not have closed properly. A resource leak can slow down or crash your system.

**Fix**
Consider closing the resource returned by the *FileResource* constructor in a try-finally block. The resource is referenced at lines 159, 171, 178. The resource is closed at line 192. There are other execution paths that do not contain closure statements, for example, when Resource.open throws an exception.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).